rest/diagnostic_api_test.go

```diff
@@ -0,0 +1,917 @@
+//  Copyright 2013-Present Couchbase, Inc.
+//
+//  Use of this software is governed by the Business Source License included
+//  in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+//  in that file, in accordance with the Business Source License, use of this
+//  software will be governed by the Apache License, Version 2.0, included in
+//  the file licenses/APL2.txt.
+
+package rest
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/couchbase/sync_gateway/auth"
+	"github.com/couchbase/sync_gateway/base"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type grant interface {
+	request(rt *RestTester)
+}
+
+const defaultKeyspace = "_default._default"
+const fakeUpdatedTime = 1234
+
+// convertUpdatedTimeToConstant replaces the updated_at field in the response with a constant value to be able to diff
+func convertUpdatedTimeToConstant(t testing.TB, output []byte) string {
+	var channelMap allChannels
+	err := json.Unmarshal(output, &channelMap)
+	require.NoError(t, err)
+	for keyspace, channels := range channelMap.Channels {
+		for channelName, grant := range channels {
+			if grant.UpdatedAt == 0 {
+				continue
+			}
+			grant.UpdatedAt = fakeUpdatedTime
+			channelMap.Channels[keyspace][channelName] = grant
+		}
+	}
+
+	fmt.Printf("Converted response: %s\n", string(base.MustJSONMarshal(t, channelMap)))
+	return string(base.MustJSONMarshal(t, channelMap))
+}
+
+func compareAllChannelsOutput(rt *RestTester, username string, expectedOutput string) {
+	response := rt.SendDiagnosticRequest(http.MethodGet,
+		"/{{.db}}/_user/"+username+"/_all_channels", ``)
+	RequireStatus(rt.TB, response, http.StatusOK)
+	rt.TB.Logf("All channels response: %s", response.BodyString())
+	require.JSONEq(rt.TB, expectedOutput, convertUpdatedTimeToConstant(rt.TB, response.BodyBytes()))
```

Comment:
I like these changes but I think we should test `_default._default` and named collections explicitly, especially because our handling of `_default._default` in user docs is different.

I think we can do this by adding golang templating to the output. I think we might be able to directly call https://github.com/couchbase/sync_gateway/blob/cef73b79270bee4ca8c85d41ac5fe9000ad367dc/rest/utilities_testing.go#L687

and replace explicit `_default._default` with `{{.keyspace}}` but you should check and let me know if this isn't true.